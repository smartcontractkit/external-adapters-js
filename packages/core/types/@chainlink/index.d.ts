// Declare missing type definitions
declare module '@chainlink/types' {
  /* REQUESTS */
  export type AdapterRequestMeta = {
    availableFunds?: number
    eligibleToSubmit?: boolean
    latestAnswer?: number
    oracleCount?: number
    paymentAmount?: number
    reportableRoundID?: number
    startedAt?: number
    timeout?: number
  }
  export type AdapterRequestDebug = {
    feedId?: string
    cacheHit?: boolean
    staleness?: number
    performance?: number
    providerCost?: number
  }
  export type AdapterRequest = {
    id: string
    data: Record<string, unknown>
    meta?: AdapterRequestMeta
    debug?: AdapterRequestDebug
    rateLimitMaxAge?: number
  }

  /* RESPONSES */
  export type DataResponse<R, P> = {
    result: R
    payload?: P
  }

  export type SequenceResponseData<R> = {
    responses?: any[]
    result: R[]
  }

  export type AdapterResponse = {
    jobRunID: string
    statusCode: number
    data: any // Response data, holds "result" for Flux Monitor. Correct way.
    result: any // Result for OCR
    maxAge?: number
    debug?: AdapterRequestDebug
  }

  /* ERRORS */
  type ErrorBasic = {
    name: string
    message: string
  }
  type ErrorFull = ErrorBasic & {
    stack: string
    cause: string
  }

  export type AdapterErrorResponse = {
    jobRunID: string
    status: string
    statusCode: number
    error: ErrorBasic | ErrorFull
  }

  /* BOOTSTRAP */
  export type Middleware = (execute: Execute, ...args: any) => Promise<Execute>
  export type Callback = (statusCode: number, data?: any) => void
  export type AdapterHealthCheck = (callback: Callback) => any

  export type { AxiosResponse, RequestConfig } from 'axios'

  export type Config = {
    apiKey?: string
    network?: string
    returnRejectedPromiseOnError?: boolean
    verbose?: boolean
    api: RequestConfig
  }

  export type ExecuteSync = (input: AdapterRequest, callback: Callback) => void

  export type Execute = (input: AdapterRequest) => Promise<AdapterResponse>

  export type ExecuteWithConfig<C extends Config> = (
    input: AdapterRequest,
    config: C,
  ) => Promise<AdapterResponse>

  export type ExecuteFactory<C extends Config> = (config?: C) => Execute

  export type ConfigFactory = (prefix?: string) => Config

  import type { ExecuteHandlers } from '@chainlink/ea-bootstrap'
  type ExecuteHandlers = ExecuteHandlers
  export type AdapterImplementation = {
    NAME: string
    makeExecute: ExecuteFactory<Config>
    makeConfig: ConfigFactory
  } & ExecuteHandlers

  /* IMPLEMENTATIONS */
  export type Address = {
    address: string
  }
  export type Account = Address & {
    balance?: string
    coin?: string
    chain?: string
    warning?: string
  }

  export type MakeWSHandler = () => WSHandler | Promise<WSHandler>
  export interface WSHandler {
    // Connection information
    connection: {
      /**
       * WS connection url
       */
      url: string
      protocol?: any
    }
    // Get the subscription message necessary to subscribe to the feed channel
    subscribe: (input: AdapterRequest) => any | undefined
    // Get unsubscribe message necessary to unsubscribe to the feed channel
    unsubscribe: (input: any) => any | undefined
    // Map to response from the incoming message and formats it into an AdapterResponse
    toResponse: (message: any) => AdapterResponse
    // Filter any message that is not from a subscribed channel
    filter: (message: any) => boolean
    // Determines if the incoming message is an error
    isError: (message: any) => boolean
    // Based on the incoming message, returns its corresponding subscription message
    subsFromMessage: (message: any) => any
  }

  /* INPUT TYPE VALIDATIONS */
  export type Override = Map<string, Map<string, string>>
}

declare module 'object-path'
declare module 'lodash'
