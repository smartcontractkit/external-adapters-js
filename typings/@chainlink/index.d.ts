// Declare missing type definitions
declare module '@chainlink/types' {
  /* REQUESTS */
  export type AdapterRequestMeta = {
    availableFunds?: number
    eligibleToSubmit?: boolean
    latestAnswer?: number
    oracleCount?: number
    paymentAmount?: number
    reportableRoundID?: number
    startedAt?: number
    timeout?: number
  }
  export type AdapterRequest = {
    id: string
    data: Record<string, unknown>
    meta?: AdapterRequestMeta
  }

  export type Callback = (statusCode: number, data?: any) => void
  export type AdapterHealthCheck = (callback: Callback) => any

  import { AxiosRequestConfig } from 'axios'
  export type Config = {
    apiKey?: string
    network?: string
    returnRejectedPromiseOnError?: Boolean
    verbose?: boolean
    api: Partial<AxiosRequestConfig>
  }

  /* RESPONSES */
  export type DataResponse<R, P> = {
    result: R
    payload?: P
  }

  export type SequenceResponseData<R> = {
    responses?: any[]
    result: R[]
  }

  export type AdapterResponse = {
    jobRunID: string
    statusCode: number
    data: any
    result: any
  }

  /* ERRORS */
  type ErrorBasic = {
    name: string
    message: string
  }
  type ErrorFull = ErrorBasic & {
    stack: string
    cause: string
  }
  export type AdapterErrorResponse = {
    jobRunID: string
    status: string
    statusCode: number
    error: ErrorBasic | ErrorFull
  }

  /* BOOTSTRAP */
  // TODO: clean this ASAP
  export type WrappedAdapterResponse = {
    statusCode: number
    data: AdapterResponse
  }
  export type ExecuteWrappedResponse = (input: AdapterRequest) => Promise<WrappedAdapterResponse>

  export type ExecuteSync = (input: AdapterRequest, callback: Callback) => void

  export type Execute = (input: AdapterRequest) => Promise<AdapterResponse>

  export type ExecuteWithConfig<C extends Config> = (
    input: AdapterRequest,
    config: C,
  ) => Promise<AdapterResponse>

  export type ExecuteFactory<C extends Config> = (config?: C) => Execute

  export type ConfigFactory = (prefix?: string) => Config

  import type { ExecuteHandlers } from '@chainlink/ea-bootstrap/src'
  export type AdapterImplementation = {
    NAME: string
    makeExecute: ExecuteFactory
    makeConfig: ConfigFactory
  } & ExecuteHandlers

  /* IMPLEMENTATIONS */
  export type Address = {
    address: string
  }
  export type Account = Address & {
    balance?: string
    coin?: string
    chain?: string
    warning?: string
  }
}
declare module '@chainlink/ea-bootstrap'
declare module '@chainlink/external-adapter'
declare module 'object-path'
