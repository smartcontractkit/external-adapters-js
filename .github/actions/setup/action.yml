name: Setup
description: Runs basic commands to bootstrap yarn and install dependencies
inputs:
  check-cache:
    description: Passes the --check-cache flag to yarn install for verification jobs
    required: false
    default: 'false'
  skip-setup:
    description: Skips the "yarn setup" step when not required
    required: false
    default: 'false'
  skip-cache:
    description: Skips cache restores and saves if set to true
    required: false
    default: 'false'
  build-all:
    description: Whether EA changes should be ignored, and all packages built regardless
    required: false
    default: 'false'
  fetch-branch:
    description: The branch that will be fetched from origin for comparison purposes
    required: false
    default: 'main'
  base-branch:
    description: The base branch to use for comparison with the head to calculate changed packages
    required: false
    default: 'origin/main'

runs:
  using: 'composite'
  steps:
    # This is a useful step that will help other actions later on
    - name: Configure git
      shell: bash
      run: |
        git config user.name "github-actions[bot]"
        git config user.email "github-actions[bot]@users.noreply.github.com"
    # First, we want to install yarn since our base image doesn't have it installed
    - uses: actions/setup-node@39370e3970a6d050c480ffad4ff0ed4d3fdee5af # v4.1.0
      with:
        # This doesn't use cache: 'yarn' as the base image doesn't have it, and we're caching node_modules
        node-version: 22
    - name: Install yarn
      run: npm i -g yarn
      shell: bash

    # Next, we want to get the dependencies, either from the cache, or fresh
    # You'll see some internet comments saying it's bad practice to cache node_modules,
    # but it should be fine in this case since we're actually looking for identical envs
    - name: Restore yarn cache
      id: restore-yarn-cache
      if: inputs.skip-cache != 'true'
      uses: actions/cache/restore@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
      with:
        # Restrict it to the run and attempt for debugging purposes, but in theory
        # these should be fine to persist between runs as well
        key: yarn_cache-${{ hashFiles('./yarn.lock') }}-${{ github.run_id }}-${{ github.run_attempt }}
        path: |
          ./.yarn/unplugged/**
          ./.yarn/install-state/**
    # If there was no cached yarn packages, install them from scratch
    - name: yarn install
      if: steps.restore-yarn-cache.outputs.cache-hit != 'true'
      run: |
        if [ "${{ inputs.check-cache }}" = "true" ]; then
          yarn install --immutable --check-cache
        else 
          yarn install --immutable
        fi
      shell: bash
    - name: Cache yarn cache
      if: steps.restore-yarn-cache.outputs.cache-hit != 'true' && inputs.skip-cache != 'true'
      uses: actions/cache/save@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
      with:
        key: yarn_cache-${{ hashFiles('./yarn.lock') }}-${{ github.run_id }}-${{ github.run_attempt }}
        path: |
          ./.yarn/unplugged/**
          ./.yarn/install-state/**

    # Build TS files unless it's been specified otherwise
    - name: Restore TS built files if present
      id: restore-ts-build
      uses: actions/cache/restore@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
      if: |
        inputs.skip-setup != 'true'
        && inputs.skip-cache != 'true'
      with:
        key: tsbuild-${{ github.sha }}-${{ github.run_id }}
        # Add the masterlist to always cache the result, even if there are no adapters to build.
        # That way we can skip the following steps in other jobs where there were no adapters built but this action is called still.
        path: |
          ./packages/**/tsconfig.tsbuildinfo
          ./packages/**/dist/
          ./MASTERLIST.md
    - name: Fetch base branch for comparison in later steps
      shell: bash
      if: |
        inputs.skip-setup != 'true'
        && steps.restore-ts-build.outputs.cache-hit != 'true'
      run: |
        git fetch origin ${{ inputs.fetch-branch }}:${{ inputs.fetch-branch }} --depth 1 --update-head-ok || true
    - name: Calculate changed EAs
      shell: bash
      if: |
        inputs.skip-setup != 'true'
        && steps.restore-ts-build.outputs.cache-hit != 'true'
      env:
        UPSTREAM_BRANCH: ${{ inputs.base-branch }}
        BUILD_ALL: ${{ inputs.build-all }}
      # We have to add the core/factories package manually always because the readme generation imports it
      run: |
        # If the BUILD_ALL env var is set or there are changes to the scripts or core (because of the readme generation scripts)
        # we need to build every possible adapter and their dependencies
        diff_output=$(git diff $UPSTREAM_BRANCH -- packages/scripts packages/core)
        if [[ $BUILD_ALL = true ]] || [[ -n $diff_output ]]; then
          cp packages/tsconfig.json packages/tsconfig.tmp.json
          exit 0
        else
          yarn workspaces list --json -R --since=$UPSTREAM_BRANCH \
          | jq -csr '.' \
          | jq '{ 
                  files: [],
                  include: [],
                  references: [
                    {
                      path: "core/factories"
                    },
                    {
                      path:  .[]
                          | .location
                          | select(
                                    startswith("packages/sources")
                                    or startswith("packages/composites")
                                    or startswith("packages/targets")
                            )
                          | match("packages/(.*)").captures[0].string
                    }
                  ]
                }' \
          > packages/tsconfig.tmp.json
        fi

        echo "CHANGED_ADAPTERS=$(jq -c '.references' packages/tsconfig.tmp.json)" >> $GITHUB_ENV
    - name: Build files
      if: |
        inputs.skip-setup != 'true'
        && steps.restore-ts-build.outputs.cache-hit != 'true'
        && env.CHANGED_ADAPTERS != '[]'
      run: |
        yarn tsc -b packages/tsconfig.tmp.json
      shell: bash
    - name: Clean tmp files
      shell: bash
      run: |
        rm -f packages/tsconfig.tmp.json
    - name: Cache built files
      if: |
        inputs.skip-setup != 'true'
        && inputs.skip-cache != 'true'
      uses: actions/cache/save@0c45773b623bea8c8e75f6c82b208c3cf94ea4f9 # v4.0.2
      with:
        key: tsbuild-${{ github.sha }}-${{ github.run_id }}
        path: |
          ./packages/**/tsconfig.tsbuildinfo
          ./packages/**/dist/
          ./MASTERLIST.md
