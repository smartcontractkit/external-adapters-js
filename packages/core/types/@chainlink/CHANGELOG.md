# @chainlink/types

## 1.1.0

### Minor Changes

- d7857c911: Created Overrider class to handle overrides & implemented Overrider class for CoinGecko EA

## 1.0.2

### Patch Changes

- 9041e0252: refactorted config into folder, moved ea presets into adapter folders, changed validator to accept ea presets

## 1.0.1

### Patch Changes

- ** DUMMY VERSION BUMP - Test release pipeline **

## 1.0.0

### Major Changes

- EAv2 Release. Start of individual EA versioning.
