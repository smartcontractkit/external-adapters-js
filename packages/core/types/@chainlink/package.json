{
  "name": "@chainlink/types",
  "version": "0.0.1",
  "types": "index.d.ts",
  "files": [
    "index.d.ts"
  ],
  "repository": {
    "url": "https://github.com/smartcontractkit/external-adapters-js",
    "type": "git"
  },
  "license": "MIT",
  "scripts": {
    "clean": "echo \"No clean needed\""
  }
}
