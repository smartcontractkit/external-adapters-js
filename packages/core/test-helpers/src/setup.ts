import { FastifyInstance } from '@chainlink/ea-bootstrap'
import request, { SuperTest, Test } from 'supertest'
import { AddressInfo } from 'net'
import * as process from 'process'
import nock from 'nock'

export type SuiteContext = {
  req: SuperTest<Test> | null
  server: () => Promise<FastifyInstance>
  fastify?: FastifyInstance
}

export type EnvVariables = { [key: string]: string }

export type TestOptions = { cleanNock?: boolean; fastify?: boolean }

export const setupExternalAdapterTest = (
  envVariables: NodeJS.ProcessEnv,
  context: SuiteContext,
  options: TestOptions = { cleanNock: true, fastify: false },
): void => {
  let fastify: FastifyInstance

  beforeAll(async () => {
    for (const key in envVariables) {
      process.env[key] = envVariables[key]
    }

    if (process.env.RECORD) {
      nock.recorder.rec()
    }
    fastify = await context.server()
    context.req = request(`localhost:${(fastify.server.address() as AddressInfo).port}`)

    // Only for edge cases when someone needs to use the fastify instance outside this function
    if (options.fastify) {
      context.fastify = fastify
    }
  })

  afterAll((done) => {
    if (process.env.RECORD) {
      nock.recorder.play()
    }

    if (options.cleanNock) {
      nock.restore()
      nock.cleanAll()
      nock.enableNetConnect()
    }

    fastify.close(done)
  })
}
