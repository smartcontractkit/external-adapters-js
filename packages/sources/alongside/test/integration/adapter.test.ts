import * as process from 'process'
import { SuperTest, Test } from 'supertest'
import { ServerInstance } from '@chainlink/external-adapter-framework'
import { mockResponseSuccess } from './fixtures'
import { SuiteContext, setupExternalAdapterTest } from './setup'

describe('rest', () => {
  jest.setTimeout(10000)
  let spy: jest.SpyInstance
  beforeAll(async () => {
    const mockDate = new Date('2022-01-01T11:11:11.111Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())
  })

  afterAll((done) => {
    spy.mockRestore()
    done()
  })

  const context: SuiteContext = {
    req: null,
    server: async () => {
      process.env['RATE_LIMIT_CAPACITY_SECOND'] = '6'
      process.env['METRICS_ENABLED'] = 'false'
      const server = (await import('../../src')).server
      return server() as Promise<ServerInstance>
    },
  }

  const envVariables = {
    ACCESS_KEY: process.env.ACCESS_KEY || 'fake-access-key',
    PASSPHRASE: process.env.PASSPHRASE || 'fake-passphrase',
    SIGNING_KEY: process.env.SIGNING_KEY || 'fake-signing-key',
    PORTFOLIO_ID: process.env.PORTFOLIO_ID || 'fake-portfolio',
    RPC_URL: process.env.RPC_URL || 'https://mainnet.infura.io:443/v3/fake-infura-key',
  }
  setupExternalAdapterTest(envVariables, context)

  describe('portfolio api', () => {
    const data = {
      data: {},
    }
    mockResponseSuccess()
    it('should return success', async () => {
      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
