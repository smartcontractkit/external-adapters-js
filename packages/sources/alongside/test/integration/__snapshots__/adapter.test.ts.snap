// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rest portfolio api should return success 1`] = `
{
  "data": {
    "result": 25954.286873113622,
  },
  "result": 25954.286873113622,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1641035471111,
    "providerDataRequestedUnixMs": 1641035471111,
  },
}
`;
